synapse/storage/databases/main/event_federation.py

```diff
@@ -1583,6 +1583,49 @@ def _record_event_failed_pull_attempt_upsert_txn(
 
         txn.execute(sql, (room_id, event_id, 1, self._clock.time_msec(), cause))
 
+    @trace
+    async def separate_event_ids_with_failed_pull_attempts(
+        self, event_ids: Collection[str]
+    ) -> Tuple[Collection[str], Collection[str]]:
+        """
+        Separate the given list of events into two lists based on whether they have any
+        failed pull attempts or not.
+
+        Args:
+            event_ids: A list of events to separate
+
+        Returns:
+            A tuple with two lists where the given event_ids are separated based on
+            whether they have any failed pull attempts or not
+            (event_ids_with_failed_pull_attempts, fresh_event_ids). Lists are ordered
+            the same as the given event_ids.
+        """
+
+        rows = await self.db_pool.simple_select_many_batch(
+            table="event_failed_pull_attempts",
+            column="event_id",
+            iterable=event_ids,
+            keyvalues={},
+            retcols=("event_id",),
+            desc="separate_event_ids_with_failed_pull_attempts",
+        )
+        event_ids_with_failed_pull_attempts_from_database = [
+            str(row["event_id"]) for row in rows
+        ]
+        # We want to maintain the order of the given event_ids so re-construct things
+        event_ids_with_failed_pull_attempts = [
+            event_id
+            for event_id in event_ids
+            if event_id in event_ids_with_failed_pull_attempts_from_database
+        ]
+        fresh_event_ids = [
+            event_id
+            for event_id in event_ids
+            if event_id not in event_ids_with_failed_pull_attempts
+        ]
```

Comment:
Instead of including this logic here, we could just move it outside and make this a little more idiomatic as `get_event_ids_with_failed_pull_attempts(...)`

Reply:
Updated to do this :fast_forward: